cli/listener/src/main.rs

```diff
@@ -210,74 +222,396 @@ pub fn get_apy_for_period(
+        let min_sleep_time = Duration::from_secs_f32(0.2);
+        let max_sleep_time = Duration::from_secs_f32(300.0);
+        let target_sleep_time = time_since_last_success.clamp(min_sleep_time, max_sleep_time);
+        let sleep_time = self
+            .rng
+            .gen_range(Duration::from_secs(0)..target_sleep_time);
+        println!("Sleeping {:?} after error ...", sleep_time);
+        sleep_time
+    }
+
+    pub fn get_sleep_time(&self) -> Duration {
+        // Sleep until is time to get the next exchange rate.
+        let sleep_time = std::time::Duration::from_secs(self.opts.poll_interval_seconds as u64);
+        println!(
+            "Sleeping for {:?} after getting the Solido exchange rate",
+            sleep_time
+        );
+        sleep_time
+    }
+}
+
+#[derive(Clone)]
+struct Metrics {
+    /// Number of times that we checked the price.
+    polls: u64,
+
+    /// Number of times that we tried to get the exchange rate, but encountered an error.
+    errors: u64,
+
+    /// Solido's maximum price interval.
+    solido_average_30d_interval_price: Option<IntervalPrices>,
+}
+
+impl Metrics {
+    pub fn write_prometheus<W: std::io::Write>(&self, out: &mut W) -> std::io::Result<()> {
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solido_pricedb_polls_total",
+                help: "Number of times we polled the exchange rate, since launch.",
+                type_: "counter",
+                metrics: vec![Metric::new(self.polls)],
+            },
+        )?;
+        write_metric(out, &MetricFamily {
+            name: "solido_pricedb_errors_total",
+            help: "Number of times we encountered an error while trying to get the exchange rate, since launch.",
+            type_: "counter",
+            metrics: vec![Metric::new(self.errors)]
+        })?;
+        if let Some(interval_price) = &self.solido_average_30d_interval_price {
+            write_metric(
+                out,
+                &MetricFamily {
+                    name: "solido_pricedb_30d_average_apy",
+                    help: "Average 30d APY",
+                    type_: "gauge",
+                    metrics: vec![Metric::new(interval_price.annual_percentage_rate())
+                        .with_label("solido", "APY".to_string())],
+                },
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+fn serve_request(request: Request, metrics_mutex: &MetricsMutex) -> Result<(), std::io::Error> {
+    // Take the current snapshot. This only holds the lock briefly, and does
+    // not prevent other threads from updating the snapshot while this request
+    // handler is running.
+    let metrics = metrics_mutex.lock().unwrap().clone();
+
+    // We don't even look at the request, for now we always serve the metrics.
+
+    let mut out: Vec<u8> = Vec::new();
+    let is_ok = metrics.write_prometheus(&mut out).is_ok();
```

Comment:
We can assume that this succeeds; we are wirting to a `Vec<u8>`. If we were writing to an `io::Write` backed by a file or a socket or something, we might get an IO error, but with a vector, the worst thing that could happen is running out of memory, but we panic in that case either way.